fuel/server.py

```diff
@@ -0,0 +1,177 @@
+        with a single `stop` key. The :func:`recv_arrays` will raise
+        ``StopIteration`` when it receives this.
+
+    See Also
+    --------
+    :meth:`zmq.Socket.send` for other arguments
+
+    Notes
+    -----
+    The protocol is very simple: A single JSON object that contains the
+    shape and data type for each array is transferred first. Subsequently
+    the arrays are sent as bytestreams (through NumPy's support of the
+    buffering protocol).
+
+    """
+    if stop:
+        mds = {'stop': True}
+        return socket.send_json(mds, flags)
+    else:
+        mds = [{'dtype': str(array.dtype), 'shape': array.shape}
+               for array in arrays]
+        socket.send_json(mds, flags | zmq.SNDMORE)
+        for array in arrays[:-1]:
+            socket.send(array, flags | zmq.SNDMORE, copy=copy, track=track)
+        return socket.send(arrays[-1], flags, copy=copy, track=track)
+
+
+def recv_arrays(socket, flags=0, copy=True, track=False):
+    """Receive a NumPy array.
+
+    Parameters
+    ----------
+    socket : :class:`zmq.Socket`
+        The socket to receive the arrays on.
+
+    Returns
+    -------
+    list
+        A list of :class:`numpy.ndarray` objects.
+
+    Raises
+    ------
+    StopIteration
+        If the first JSON object received contains the key `stop`.
+
+    See Also
+    --------
+    :func:`send_arrays` for an explanation of the protocol used to transfer
+    arrays, and :meth:`zmq.Socket.recv` for an explanation of the other
+    arguments.
+
+    """
+    mds = socket.recv_json(flags=flags)
+    if 'stop' in mds:
+        raise StopIteration
+    arrays = []
+    for md in mds:
+        data = socket.recv(flags=flags, copy=copy, track=track)
+        buf = buffer_(data)
+        array = numpy.frombuffer(buf, dtype=md['dtype'])
+        arrays.append(array.reshape(md['shape']))
+    return arrays
+
+
+def server(data_stream, server_port):
+    """The main process that processes and sends data."""
+    context = zmq.Context()
+    socket = context.socket(zmq.REP)
+    socket.connect("tcp://localhost:{}".format(server_port))
+
+    it = data_stream.get_epoch_iterator()
+
+    logger.info('server started')
+    while True:
+        socket.recv()
+        try:
+            data = next(it)
+            stop = False
+            logger.info("sending {} arrays".format(len(data)))
+        except StopIteration:
+            it = data_stream.get_epoch_iterator()
+            data = None
+            stop = True
+            logger.info("sending StopIteration")
+        send_arrays(socket, data, stop=stop)
+
+
+def broker(server_port, client_port):
+    """The broker is run in a separate process and holds the queue."""
+    logger.debug('binding broker to sockets')
+    context = zmq.Context()
+    frontend = context.socket(zmq.ROUTER)
+    backend = context.socket(zmq.DEALER)
+    frontend.bind("tcp://*:{}".format(client_port))
+    backend.bind("tcp://*:{}".format(server_port))
```

Comment:
Is tcp what we want to be using always? what about unix sockets? (I have no idea)

Also, what exactly does the * entail? (Will look into this after dinner if you haven't gotten back)